### PR TITLE
feature: support kafka schema registry for Protobuf and Avro.

### PR DESCRIPTION
**Pull request by zliang-min**
PR checklist:
- Did you run ClangFormat ?
- Did you separate headers to a different section in existing community code base ?
- Did you surround `proton: starts/ends` for new code in existing community code base ?

Please write user-readable short description of the changes:

New settings added for using kafka schema registry: `kafka_schema_registry_url` and `kafka_schema_registry_credentials`. Example:

```sql
CREATE EXTERNAL STREAM my_stream (
  -- columns goes here ...
) SETTINGS
type = 'kafka',
brokers = '...',
topic = '...',
data_format = 'ProtobufSingle',
kafka_schema_registry_url = 'http://url.to/my/schema/registry',
kafka_schema_registry_credentials = '$api_key}:${secret}';
```

`kafka_schema_registry_credentials` is optional, it contains the the API key and secret and a `:` in between. Also, kafka schema registry can be used with `data_format = 'Avro'`.

Note: in this PR, schema reference is not supported yet.

### CODE REVIEW

File: src/Formats/FormatSchemaFactory.cpp [hunk 1]
```diff
@@ -4,6 +4,10 @@
 #include <IO/ReadBufferFromString.h>
 #include <Processors/Formats/ISchemaWriter.h>
 
+/// proton: starts
```
Comment (chenziliang) [inline src/Formats/FormatSchemaFactory.cpp]: FormatSchema is added by us ? If yes, we can skip the start/end comment

File: src/Formats/KafkaSchemaRegistry.h [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+namespace DB
+{
+
+/// A helper class helps working with Kafka schema registry.
+class KafkaSchemaRegistry final
+{
+public:
+    static UInt32 readSchemaId(ReadBuffer & in);
+
+    /// `credentials_` is expected to be formatted in "<username>:<password>".
```
Comment (chenziliang) [inline src/Formats/KafkaSchemaRegistry.h]: `/// \param credentials_ it is expected to be ...`

File: src/Formats/KafkaSchemaRegistry.cpp [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+extern const int INCORRECT_DATA;
+}
+
+KafkaSchemaRegistry::KafkaSchemaRegistry(const String & base_url_, const String & credentials_): base_url(base_url_)
+{
+    if (credentials_.empty())
+        return;
+
+    auto pos = credentials_.find(':');
+    if (pos == credentials_.npos)
```
Comment (chenziliang) [inline src/Formats/KafkaSchemaRegistry.cpp]: we can do 

```
if (auto pos = credentials_.find(':'); pos == credentials_.npos)
{
}
else
{
}
```

Or 

```
if (auto pos = credentials_.find(':'); pos != credentials_.npos)
{
    credentials.setUsername(credentials_.substr(0, pos));
    credentials.setPassword(credentials_.substr(pos + 1));
}
else
{
    credentials.setUsername(credentials_);
}
```

File: src/Formats/KafkaSchemaRegistry.cpp [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+{
+
+namespace ErrorCodes
+{
+extern const int INCORRECT_DATA;
+}
+
+KafkaSchemaRegistry::KafkaSchemaRegistry(const String & base_url_, const String & credentials_): base_url(base_url_)
+{
+    if (credentials_.empty())
```
Comment (chenziliang) [inline src/Formats/KafkaSchemaRegistry.cpp]: Probably we don't need this test since setting an empty username shall be no-op

File: src/Formats/KafkaSchemaRegistry.cpp [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    else
+    {
+        credentials.setUsername(credentials_.substr(0, pos));
+        credentials.setPassword(credentials_.substr(pos + 1));
+    }
+}
+
+String KafkaSchemaRegistry::fetchSchema(UInt32 id)
+{
+    assert(!base_url.empty());
```
Comment (chenziliang) [inline src/Formats/KafkaSchemaRegistry.cpp]: we can assert in ctor since this is the only place we got it inited

File: src/Formats/KafkaSchemaRegistry.cpp [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+
+    try
+    {
+        try
+        {
+            Poco::URI url(base_url, "/schemas/ids/" + std::to_string(id));
```
Comment (chenziliang) [inline src/Formats/KafkaSchemaRegistry.cpp]: `fmt::format("/schemas/ids/{}", id)` shall be do the work and do it pretty good ?

File: src/Formats/KafkaSchemaRegistry.cpp [hunk 7]
```diff
@@ -0,0 +1,119 @@
+#include <Formats/KafkaSchemaRegistry.h>
+#include <IO/HTTPCommon.h>
+#include <IO/ReadHelpers.h>
+#include <Poco/JSON/Parser.h>
```
Comment (chenziliang) [inline src/Formats/KafkaSchemaRegistry.cpp:6]: Let's separate 3rd party libs includes to a new include segment

File: src/Formats/KafkaSchemaRegistry.cpp [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+    {
+        readBinaryBigEndian(magic, in);
+        readBinaryBigEndian(schema_id, in);
+    }
+    catch (const Exception & e)
+    {
+        if (e.code() == ErrorCodes::CANNOT_READ_ALL_DATA)
+        {
+            /// empty or incomplete message without magic byte or schema id
+            throw Exception("Missing magic byte or schema identifier.", ErrorCodes::INCORRECT_DATA);
```
Comment (chenziliang) [inline src/Formats/KafkaSchemaRegistry.cpp]: Let's use this format of exception construction

```
throw Exception(error_code, fmt_string, fmt_args);
```

File: src/Formats/KafkaSchemaRegistry.cpp [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+        }
+        else
+            throw;
+    }
+
+    if (magic != 0x00)
+    {
+        throw Exception("Invalid magic byte before schema identifier."
```
Comment (chenziliang) [inline src/Formats/KafkaSchemaRegistry.cpp]: Same here

```
throw Exception(ErrorCodes::INCORRECT_DATA, "Invalid magic byte before schema identifier. Must be zero byte, found 0x{:x} instead", magic);
```

File: src/Formats/ProtobufSchemas.h [hunk 10]
```diff
@@ -21,28 +25,28 @@ namespace protobuf
...
-/// proton: starts
-struct SchemaValidationError;
-using SchemaValidationErrors = std::vector<SchemaValidationError>;
-/// proton: ends
 
 /** Keeps parsed google protobuf schemas parsed from files.
   * This class is used to handle the "Protobuf" input/output formats.
   */
-class ProtobufSchemas : private boost::noncopyable
+class ProtobufSchemas : private boost::noncopyable, public google::protobuf::io::ErrorCollector /* proton: updated */
```
Comment (chenziliang) [inline src/Formats/ProtobufSchemas.h]: We can remove `private boost::noncopyable` probably since the data member std::mutex is not copyable which effectively make ProtobufSchemas non-copyable.

File: src/Formats/ProtobufSchemas.cpp [hunk 11]
```diff
@@ -106,17 +83,24 @@ const google::protobuf::Descriptor * ProtobufSchemas::getMessageTypeForFormatSch
 }
 
 /// proton: starts
-SchemaValidationErrors ProtobufSchemas::validateSchema(std::string_view schema)
+/// Overrides google::protobuf::io::ErrorCollector:
+void ProtobufSchemas::AddError(int line, google::protobuf::io::ColumnNumber column, const std::string & message)
+{
+    throw Exception(
```
Comment (chenziliang) [inline src/Formats/ProtobufSchemas.cpp]: Same exception construction comment as above

File: src/Processors/Formats/ISchemaWriter.h [hunk 12]
```diff
@@ -30,8 +16,8 @@ class IExternalSchemaWriter
 
     virtual ~IExternalSchemaWriter() = default;
 
-    /// Validates the schema input.
-    virtual SchemaValidationErrors validate() = 0;
+    /// Validates the schema input. Should throws exceptions on validation failures.
```
Comment (chenziliang) [inline src/Processors/Formats/ISchemaWriter.h]: `Should throw`

File: src/Processors/Formats/Impl/AvroRowInputFormat.cpp [hunk 13]
```diff
@@ -663,64 +752,23 @@ class AvroConfluentRowInputFormat::SchemaRegistry
...
-            }
+            return avro::compileJsonSchemaFromString(schema);
         }
-        catch (Exception & e)
+        catch (const avro::Exception & e)
         {
-            e.addMessage("while fetching schema id = " + std::to_string(id));
-            throw;
+          auto ex = Exception(e.what(), ErrorCodes::INCORRECT_DATA);
+          ex.addMessage("while fetching schema id = " + std::to_string(id));
```
Comment (chenziliang) [inline src/Processors/Formats/Impl/AvroRowInputFormat.cpp]: we can consider `fmt::format(...)` here if it works

File: src/Processors/Formats/Impl/AvroRowInputFormat.cpp [hunk 14]
```diff
@@ -663,64 +752,23 @@ class AvroConfluentRowInputFormat::SchemaRegistry
...
+          throw std::move(ex);
```
Comment (chenziliang) [inline src/Processors/Formats/Impl/AvroRowInputFormat.cpp:685]: `throw ex` is good

File: src/Formats/KafkaSchemaRegistry.cpp [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+            throw;
+        }
+        catch (const Poco::Exception & e)
+        {
+            throw Exception(Exception::CreateFromPocoTag{}, e);
+        }
+    }
+    catch (Exception & e)
+    {
+        e.addMessage("while fetching schema id = " + std::to_string(id));
```
Comment (chenziliang) [inline src/Formats/KafkaSchemaRegistry.cpp]: we can consider using `fmt::format(...)` here

File: src/Formats/KafkaSchemaRegistry.cpp [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+String KafkaSchemaRegistry::fetchSchema(UInt32 id)
+{
+    assert(!base_url.empty());
+
+    try
+    {
+        try
+        {
+            Poco::URI url(base_url, "/schemas/ids/" + std::to_string(id));
+            LOG_TRACE((&Poco::Logger::get("KafkaSchemaRegistry")), "Fetching schema id = {}", id);
```
Comment (chenziliang) [inline src/Formats/KafkaSchemaRegistry.cpp]: We can consider cache the `logger` in the registry class since it is used twice already

File: src/Processors/Formats/Impl/AvroRowInputFormat.cpp [hunk 17]
```diff
@@ -729,57 +777,62 @@ static LRUCache<std::string, ConfluentSchemaRegistry>  schema_registry_cache(SCH
 
 static std::shared_ptr<ConfluentSchemaRegistry> getConfluentSchemaRegistry(const FormatSettings & format_settings)
 {
-    const auto & base_url = format_settings.avro.schema_registry_url;
+    /// proton: starts
+    const auto & base_url = format_settings.schema.kafka_schema_registry_url.empty() ? format_settings.avro.schema_registry_url : format_settings.schema.kafka_schema_registry_url;
```
Comment (chenziliang) [inline src/Processors/Formats/Impl/AvroRowInputFormat.cpp:702]: We are using different urls for avro and kafka but share the same credential : `kafka_schema_registry_credentials` ? Or for avro, `kafka_schema_registry_credentials` will be empty always ?

Comment (chenziliang) [reply src/Processors/Formats/Impl/AvroRowInputFormat.cpp:702]: Do we need `kafka_schema_registry_url` ? or we can just consolidate them like existing code does ?

File: src/Processors/Formats/Impl/AvroRowInputFormat.cpp [hunk 18]
```diff
@@ -729,57 +777,62 @@ static LRUCache<std::string, ConfluentSchemaRegistry>  schema_registry_cache(SCH
...
+    const auto & credentials = format_settings.schema.kafka_schema_registry_credentials;
     auto [schema_registry, loaded] = schema_registry_cache.getOrSet(
-        base_url,
-        [base_url]()
+        base_url + credentials,
+        [base_url, credentials]()
```
Comment (chenziliang) [inline src/Processors/Formats/Impl/AvroRowInputFormat.cpp]: if this is a sync call, we can ref capture here for base_url and credentials ? like 

```
[&] {
}
```

File: src/Processors/Formats/Impl/AvroRowInputFormat.h [hunk 19]
```diff
@@ -155,7 +155,7 @@ class AvroConfluentRowInputFormat final : public IRowInputFormat
     std::unordered_map<SchemaId, AvroDeserializer> deserializer_cache;
     const AvroDeserializer & getOrCreateDeserializer(SchemaId schema_id);
 
-    avro::InputStreamPtr input_stream;
+    /// avro::InputStreamPtr input_stream; /* proton: updated */
```
Comment (chenziliang) [inline src/Processors/Formats/Impl/AvroRowInputFormat.h]: I am thinking eliminating them all for deletion since we know what we are doing to do and we are doing better than CH community.

File: src/Processors/Formats/Impl/AvroRowInputFormat.cpp [hunk 20]
```diff
@@ -910,9 +969,18 @@ void registerInputFormatAvro(FormatFactory & factory)
         ReadBuffer & buf,
         const Block & sample,
         const RowInputFormatParams & params,
-        const FormatSettings & settings)
+        const FormatSettings & settings) -> InputFormatPtr
     {
-        return std::make_shared<AvroRowInputFormat>(sample, buf, params, settings);
+        /// proton: starts
+        /// Use only one format name "Avro" to support both shema registry and non-schema registry use cases, rather than using another name "AvroConfluent"
```
Comment (chenziliang) [inline src/Processors/Formats/Impl/AvroRowInputFormat.cpp]: `both schema registry ...`

File: src/Processors/Formats/Impl/AvroRowInputFormat.cpp [hunk 21]
```diff
@@ -910,9 +969,18 @@ void registerInputFormatAvro(FormatFactory & factory)
...
+
+        if (settings.avro.schema_registry_url.empty() && settings.schema.kafka_schema_registry_url.empty())
+            return std::make_shared<AvroRowInputFormat>(sample, buf, params, settings);
```
Comment (chenziliang) [inline src/Processors/Formats/Impl/AvroRowInputFormat.cpp:867]: Let's comment `/// Non-schema registry case`

File: src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp [hunk 22]
```diff
@@ -66,21 +83,43 @@ void ProtobufRowInputFormat::syncAfterError()
...
+        [](ReadBuffer & buf, const Block & sample, IRowInputFormat::Params params, const FormatSettings & settings) -> std::shared_ptr<IInputFormat>
         {
-            return std::make_shared<ProtobufRowInputFormat>(buf, sample, std::move(params),
-                FormatSchemaInfo(settings, "Protobuf", true),
-                with_length_delimiter);
+            if (settings.schema.kafka_schema_registry_url.empty())
+                return std::make_shared<ProtobufRowInputFormat>(
+                    buf, sample, std::move(params), FormatSchemaInfo(settings, "Protobuf", true), /*with_length_delimiter=*/false);
+
+            if (!settings.schema.format_schema.empty())
```
Comment (chenziliang) [inline src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp:111]: Since we have done the validation here, do we still need do validation in ProtobufRowInputFormat or  ProtobufConfluentRowInputFormat ? Can we just assert there instead ?

File: src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp [hunk 23]
```diff
@@ -66,21 +83,43 @@ void ProtobufRowInputFormat::syncAfterError()
...
+    factory.registerInputFormat(
+        "ProtobufSingle",
+        [](ReadBuffer & buf, const Block & sample, IRowInputFormat::Params params, const FormatSettings & settings) -> std::shared_ptr<IInputFormat>
         {
-            return std::make_shared<ProtobufRowInputFormat>(buf, sample, std::move(params),
-                FormatSchemaInfo(settings, "Protobuf", true),
-                with_length_delimiter);
+            if (settings.schema.kafka_schema_registry_url.empty())
+                return std::make_shared<ProtobufRowInputFormat>(
+                    buf, sample, std::move(params), FormatSchemaInfo(settings, "Protobuf", true), /*with_length_delimiter=*/false);
```
Comment (chenziliang) [inline src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp:109]: So we are using `Protobuf` instead of `ProtobufSingle` for length_delimiter = false case ?

File: src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp [hunk 24]
```diff
@@ -100,20 +139,162 @@ NamesAndTypesList ProtobufSchemaReader::readSchema()
 }
 
 /// proton: starts
+namespace
+{
+using ConfluentSchemaRegistry = ProtobufConfluentRowInputFormat::SchemaRegistryWithCache;
+#define SCHEMA_REGISTRY_CACHE_MAX_SIZE 1000
+/// Cache of Schema Registry URL + credentials -> SchemaRegistry
+LRUCache<std::string, ConfluentSchemaRegistry>  schema_registry_cache(SCHEMA_REGISTRY_CACHE_MAX_SIZE);
```
Comment (chenziliang) [inline src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp]: Can we wrap this in a function like 

```
auto & schemaRegitryCache()
{
    using ConfluentSchemaRegistry = ProtobufConfluentRowInputFormat::SchemaRegistryWithCache;

    /// Cache of Schema Registry URL + credentials -> SchemaRegistry
    static LRUCache<std::string, ConfluentSchemaRegistry>  schema_registry_cache(/*max_size_=*/1000);
    return schema_registry_cache;
}
```

File: src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp [hunk 25]
```diff
@@ -100,20 +139,162 @@ NamesAndTypesList ProtobufSchemaReader::readSchema()
...
+
+std::shared_ptr<ConfluentSchemaRegistry> getConfluentSchemaRegistry(const String & base_url, const String & credentials)
+{
+    auto [schema_registry, loaded] = schema_registry_cache.getOrSet(
+        base_url + credentials,
+        [base_url, credentials]()
```
Comment (chenziliang) [inline src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp]: ref capture

File: src/Processors/Formats/Impl/ProtobufRowInputFormat.h [hunk 26]
```diff
@@ -7,6 +7,9 @@
 #    include <Processors/Formats/IRowInputFormat.h>
 #    include <Processors/Formats/ISchemaReader.h>
 #    include <Processors/Formats/ISchemaWriter.h>
+/// proton: starts
```
Comment (chenziliang) [inline src/Processors/Formats/Impl/ProtobufRowInputFormat.h:11]: Let's insert an empty line before this segment

File: src/Processors/Formats/Impl/ProtobufRowInputFormat.h [hunk 27]
```diff
@@ -61,12 +69,36 @@ class ProtobufSchemaReader : public IExternalSchemaReader
 };
 
 /// proton: starts
+/// Confluent framing + Protobuf binary datum encoding. Mainly used for Kafka.
+/// Uses 3 caches:
+/// 1. global: schema registry cache (base_url -> SchemaRegistry)
```
Comment (chenziliang) [inline src/Processors/Formats/Impl/ProtobufRowInputFormat.h]: `(base_url + credentials -> SchemaRegistry)` ?

File: src/Processors/Formats/Impl/ProtobufRowInputFormat.h [hunk 28]
```diff
@@ -61,12 +69,36 @@ class ProtobufSchemaReader : public IExternalSchemaReader
...
+/// 2. SchemaRegistry: schema cache (schema_id -> schema)
+/// 3. ProtobufConfluentRowInputFormat: deserializer cache (schema_id -> AvroDeserializer)
+/// This is needed because KafkaStorage creates a new instance of InputFormat per a batch of messages
```
Comment (chenziliang) [inline src/Processors/Formats/Impl/ProtobufRowInputFormat.h:78]: Wondering how do we do cache invalidation ? do we support it now and how it works in prod for schema evolution / changes ?

File: src/Processors/Formats/Impl/ProtobufRowInputFormat.h [hunk 29]
```diff
@@ -61,12 +69,36 @@ class ProtobufSchemaReader : public IExternalSchemaReader
...
+class ProtobufConfluentRowInputFormat final : public IRowInputFormat
+{
+public:
+    ProtobufConfluentRowInputFormat(ReadBuffer & in_, const Block & header_, Params params_, const FormatSettings & format_settings_);
+    String getName() const override { return "ProtobufConfluentRowInputFormat"; }
+
+    // void setReadBuffer(ReadBuffer & buf) override;
+
+    class SchemaRegistryWithCache;
```
Comment (chenziliang) [inline src/Processors/Formats/Impl/ProtobufRowInputFormat.h:87]: We can forward declare before the class definition ?

Comment (chenziliang) [reply src/Processors/Formats/Impl/ProtobufRowInputFormat.h:87]: Ok, it is expected to be an inner class

File: src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp [hunk 30]
```diff
@@ -100,20 +139,162 @@ NamesAndTypesList ProtobufSchemaReader::readSchema()
...
+
+        auto const * ret = registry_pool()->BuildFile(descriptor);
+        auto mt_count = ret->message_type_count();
+        if (mt_count < 1)
+            throw Exception(ErrorCodes::INVALID_DATA, "No message type in schema");
+        return ret;
+    }
+
+    KafkaSchemaRegistry registry;
+    google::protobuf::DescriptorPool registry_pool_;
```
Comment (chenziliang) [inline src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp]: `registry_pool`

File: src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp [hunk 31]
```diff
@@ -100,20 +139,162 @@ NamesAndTypesList ProtobufSchemaReader::readSchema()
...
+
+    google::protobuf::DescriptorPool* registry_pool() { return &registry_pool_; }
```
Comment (chenziliang) [inline src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp]: `auto * regstryPool() noexcept` and do we need it ?

File: src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp (hunk 30 above)
Comment (chenziliang) [reply src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp]: I feel like it shall be `descriptor_pool` ?

File: src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp [hunk 32]
```diff
@@ -100,20 +139,162 @@ NamesAndTypesList ProtobufSchemaReader::readSchema()
...
+    const google::protobuf::FileDescriptor * getSchema(uint32_t id)
+    {
+        const auto * loaded_descriptor = registry_pool()->FindFileByName(std::to_string(id));
+        if (loaded_descriptor)
+            return loaded_descriptor;
+
+        return fetchSchema(id);
+    }
+
+    const google::protobuf::FileDescriptor* fetchSchema(uint32_t id)
```
Comment (chenziliang) [inline src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp]: We will need some clang-format for the code we added

File: src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp [hunk 33]
```diff
@@ -100,20 +139,162 @@ NamesAndTypesList ProtobufSchemaReader::readSchema()
...
+            descriptor = descriptor->nested_type(i);
+        }
+
+        return descriptor;
+    }
+
+private:
+    const google::protobuf::FileDescriptor * getSchema(uint32_t id)
+    {
+        const auto * loaded_descriptor = registry_pool()->FindFileByName(std::to_string(id));
```
Comment (chenziliang) [inline src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp]: Do we need call `registry_pool()` function ?

```
if (const auto * loaded_descriptor = descriptor_pool.FindFileByName(std::to_string(id)); loaded_descriptor != nullptr)
    return loaded_descriptor;
```

Comment (chenziliang) [reply src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp]: Does `id` uniquely identify the descriptor ?

File: src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp [hunk 34]
```diff
@@ -100,20 +139,162 @@ NamesAndTypesList ProtobufSchemaReader::readSchema()
...
+        if (loaded_descriptor)
+            return loaded_descriptor;
+
+        return fetchSchema(id);
+    }
+
+    const google::protobuf::FileDescriptor* fetchSchema(uint32_t id)
+    {
+        auto schema = registry.fetchSchema(id);
+        std::string schema_content {schema.data(), schema.size()};
```
Comment (chenziliang) [inline src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp]: since `registry.fetchSchema(id)` already returns a full copy of a schema string, do we need do another copy (`schema_content`) ?

File: src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp [hunk 35]
```diff
@@ -100,20 +139,162 @@ NamesAndTypesList ProtobufSchemaReader::readSchema()
...
+        google::protobuf::io::ArrayInputStream input{schema.data(), static_cast<int>(schema.size())};
+        google::protobuf::io::Tokenizer tokenizer(&input, this);
+        google::protobuf::FileDescriptorProto descriptor;
+        descriptor.set_name(std::to_string(id));
+        google::protobuf::compiler::Parser parser;
+        parser.RecordErrorsTo(this);
+        parser.Parse(&tokenizer, &descriptor);
+
+        auto const * ret = registry_pool()->BuildFile(descriptor);
```
Comment (chenziliang) [inline src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp]: We will need check nullness of ret

```
if (const auto * descriptor = descriptor_pool.BuildFile(descriptor_proto); descriptor && descriptor->message_type_count() >= 1)
   return descriptor;

throw ...
```

File: src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp [hunk 36]
```diff
@@ -100,20 +139,162 @@ NamesAndTypesList ProtobufSchemaReader::readSchema()
...
+    auto schema_id = KafkaSchemaRegistry::readSchemaId(*in);
+
+
+    Int64 indexes_count = 0;
+    readVarInt(indexes_count, *in);
+
+    std::vector<Int64> indexes;
+    indexes.reserve(indexes_count < 1 ? 1 : indexes_count);
+
+    if (indexes_count < 1)
```
Comment (chenziliang) [inline src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp:265]: nit : Let's put a curly braces here for the `if` clause

File: src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp [hunk 37]
```diff
@@ -100,20 +139,162 @@ NamesAndTypesList ProtobufSchemaReader::readSchema()
...
+        indexes.push_back(0);
+    else
+    {
+        for (size_t i = 0; i < static_cast<size_t>(indexes_count); i++)
```
Comment (chenziliang) [inline src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp]: `for (Int64_t i = 0; ...`

File: src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp [hunk 38]
```diff
@@ -100,20 +139,162 @@ NamesAndTypesList ProtobufSchemaReader::readSchema()
...
+    ProtobufReader reader {*in};
+    serializer = ProtobufSerializer::create(
+        header.getNames(),
+        header.getDataTypes(),
+        missing_column_indices,
+        *registry->getMessageType(schema_id, indexes),
+        /*with_length_delimiter=*/false,
+        reader);
+
+    size_t row_num = columns.empty() ? 0 : columns[0]->size();
```
Comment (chenziliang) [inline src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp:290]: is it possible we have empty columns to read ?

Comment (chenziliang) [reply src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp:290]: i c, `columns` are grown automatically when we read, and initially no rows in the columns

File: src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp [hunk 39]
```diff
@@ -100,20 +139,162 @@ NamesAndTypesList ProtobufSchemaReader::readSchema()
...
+    void AddError(int line, google::protobuf::io::ColumnNumber column, const std::string & message) override
+    {
+        throw Exception(ErrorCodes::INVALID_DATA, "Failed to parse schema, line={}, column={}, message={}", line, column, message);
+    }
+
+    const google::protobuf::Descriptor * getMessageType(uint32_t schema_id, const std::vector<Int64> & indexes)
+    {
+        assert(!indexes.empty());
+
+        const auto *fd = getSchema(schema_id);
```
Comment (chenziliang) [inline src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp]: clang-format

File: src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp [hunk 40]
```diff
@@ -100,20 +139,162 @@ NamesAndTypesList ProtobufSchemaReader::readSchema()
...
+        auto mt_count = fd->message_type_count();
+        if (mt_count < indexes[0] + 1)
+            throw Exception(ErrorCodes::INVALID_DATA, "Invalid message index={} max_index={}", indexes[0], mt_count);
+
+        const auto *descriptor = fd->message_type(indexes[0]);
```
Comment (chenziliang) [inline src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp]: clang-format

File: src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp [hunk 41]
```diff
@@ -100,20 +139,162 @@ NamesAndTypesList ProtobufSchemaReader::readSchema()
...
+
+        for (auto i : std::vector<Int64>(indexes.begin() + 1, indexes.end()))
```
Comment (chenziliang) [inline src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp]: Let's use std::span in `std::ranges` which avoids a copy

Comment (chenziliang) [inline src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp]: We may need put some comments here to explain the loop. It appears to me, we start from the initial index and follow the chain to get the last descriptor

File: src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp [hunk 42]
```diff
@@ -100,20 +139,162 @@ NamesAndTypesList ProtobufSchemaReader::readSchema()
...
+        reader);
+
+    size_t row_num = columns.empty() ? 0 : columns[0]->size();
+    if (!row_num)
+        serializer->setColumns(columns.data(), columns.size());
+
+    serializer->readRow(row_num);
+
+    row_read_extension.read_columns.clear();
+    row_read_extension.read_columns.resize(columns.size(), true);
```
Comment (chenziliang) [inline src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp]: if missing_column_indices is zero, i was thinking we can optimize this as  following ? and i recalled the framework (parent driving template function) does the clear before the call

```
assert(row_read_extension.read_columns.empty());
if (!missing_column_indices.empty())
{
    row_read_extension.read_columns.resize(columns.size(), true);
     for (size_t column_idx : missing_column_indices)
        row_read_extension.read_columns[column_idx] = false;
}

```

File: src/Formats/FormatSchemaFactory.cpp (hunk 1 above)
Comment (zliang-min) [reply src/Formats/FormatSchemaFactory.cpp]: Oops, that's right.

**Comment by chenziliang** [comment]
In general, it looks good and let's clean up the old code which is commented out

File: src/Processors/Formats/Impl/AvroRowInputFormat.cpp (hunk 14 above)
Comment (zliang-min) [reply src/Processors/Formats/Impl/AvroRowInputFormat.cpp:685]: `throw ex` reports `Throw expression should throw anonymous temporary values instead`.

File: src/Processors/Formats/Impl/AvroRowInputFormat.cpp (hunk 17 above)
Comment (zliang-min) [reply src/Processors/Formats/Impl/AvroRowInputFormat.cpp:702]: Actually we probably should remove `avro.schema_registry_url` eventually. I don't know why it was designed in that way, because schema registry is not Avro specific. `kafka_schema_registry` is more generic, and it can be used in Protobuf, Avro, and JSON Schema (if we want to support in the future).

`kafka_schema_registry_credentials` is needed for authentication. The previous Avro implementation does not support authentication, which means it could not work with Conflent cloud.

File: src/Processors/Formats/Impl/AvroRowInputFormat.cpp (hunk 18 above)
Comment (zliang-min) [reply src/Processors/Formats/Impl/AvroRowInputFormat.cpp]: Yes, I think it should be fine, because eventually, `KafkaSchemaREgistry` will copy the URL and credentials.

File: src/Processors/Formats/Impl/AvroRowInputFormat.h (hunk 19 above)
Comment (zliang-min) [reply src/Processors/Formats/Impl/AvroRowInputFormat.h]: Yeah, I think that should be fine, because these code are quite specific and reasonably easy to understand if even we want to pick up some changes from CH in the future.

File: src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp (hunk 23 above)
Comment (zliang-min) [reply src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp:109]: No, I didn't change that in this PR. It's still `ProtobufSingle`. It's that for `FormatSchemaInfo` they all belong to `"Protobuf"`.

File: src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp (hunk 22 above)
Comment (zliang-min) [reply src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp:111]: Neither ProtobufRowInputFormat nor ProtobufConfluentRowInputFormat duplicates this validation, here is the only place does it.

File: src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp (hunk 41 above)
Comment (zliang-min) [reply src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp]: I put a comment there and put a link to the explanation.

File: src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp (hunk 33 above)
Comment (zliang-min) [reply src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp]: Yes, schema IDs are unique, each one points to a specific version of a schema.

File: src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp (hunk 34 above)
Comment (zliang-min) [reply src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp]: That's a useless code being forgotten to be deleted.

File: src/Processors/Formats/Impl/ProtobufRowInputFormat.h (hunk 28 above)
Comment (zliang-min) [reply src/Processors/Formats/Impl/ProtobufRowInputFormat.h:78]: I think we don't need to worry about the invalidation much for now. Schemas (on a schema registry) are immutable, i.e. someone updates a schema (also called evolves the schema), a new one will be created with a new schema ID, so in fact, the existing cache is still valid ( one can still use the old schema to write data, it's toally valid ).

File: src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp (hunk 42 above)
Comment (zliang-min) [reply src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp]: lI applied your version, and it works fine in my tests, so I think this is fine.

File: src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp (hunk 31 above)
Comment (zliang-min) [reply src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp]: No, it's actually not needed (tested), I believe it's a left-over after some refactor.

**Review by chenziliang — APPROVED** [review]
